devicetypes/plaidsystems/spruce-sensor.src/spruce-sensor.groovy

```diff
@@ -12,406 +12,259 @@
-    }
- 	log.debug "result: $result"
-    return result
-    
-}
+	else if (isSupportedDescription(description)) {
+		log.debug "supported description: $description"
+		map = parseSupportedMessage(description)
+	}
+	else if (description?.startsWith("catchall:")) {
+		log.debug "catchall ${description}"
+		map = parseCatchAllMessage(description)
+	}
+	else if (DEBUG) log.debug "uncaught ${description}"
 
+	def result = map ? createEvent(map) : null
 
+	//check for configuration change and send configuration change
+	if (map && map.name == "temperature" && isIntervalChange()) result = ping()
+
+	if (DEBUG) log.debug "parse result: $result"
+	return result
+}
 
 private Map parseCatchAllMessage(String description) {
-    Map resultMap = [:]
-    def linkText = getLinkText(device)
-	//log.debug "Catchall"
-    def descMap = zigbee.parse(description)
-    
-    //check humidity configuration is complete
-    if (descMap.command == 0x07 && descMap.clusterId == 0x0405){    	
-        def configInterval = 10
-        if (interval != null) configInterval = interval        
-        sendEvent(name: 'configuration',value: configInterval, descriptionText: "Configuration Successful")        
-        //setConfig()
-        log.debug "config complete"        
-        //return resultMap = [name: 'configuration', value: configInterval, descriptionText: "Settings configured successfully"]
-    }
-    else if (descMap.command == 0x0001){    
-    	def hexString = "${hex(descMap.data[5])}" + "${hex(descMap.data[4])}"
-    	def intString = Integer.parseInt(hexString, 16)    
-    	//log.debug "command: $descMap.command clusterid: $descMap.clusterId $hexString $intString"
-    
-    	if (descMap.clusterId == 0x0402){    	
-            def value = getTemperature(hexString)
-            resultMap = getTemperatureResult(value)    
-        }
-        else if (descMap.clusterId == 0x0405){
-            def value = Math.round(new BigDecimal(intString / 100)).toString()
-            resultMap = getHumidityResult(value)
-
-        }
-        else return null
-    }
-    else return null 
-    
-    return resultMap
-}    
-    
-private Map parseReportAttributeMessage(String description) {	
-    def descMap = parseDescriptionAsMap(description)
-	log.debug "Desc Map: $descMap"
-    log.debug "Report Attributes"
- 
 	Map resultMap = [:]
-	if (descMap.cluster == "0001" && descMap.attrId == "0000") {		
-        resultMap = getBatteryResult(descMap.value)
-	}    
-    return resultMap
+	def map = zigbee.parseDescriptionAsMap(description)
+
+	def command = zigbee.convertHexToInt(map.command)
+	def cluster = ( map.clusterId == null ? zigbee.convertHexToInt(map.cluster) : zigbee.convertHexToInt(map.clusterId) )
+	def value = (map.value != null ? zigbee.convertHexToInt(map.value) : null)
+
+	if (DEBUG) log.debug "command: ${command} cluster: ${cluster} value: ${value}"
+
+	//check humidity configuration update is complete
+	if (command == 0x07 && cluster == 0x0405){
```

Comment:
Please change all "magic numbers" to constants, like:
`private getBATTERY_PERCENTAGE_REMAINING() { 0x0021 }`
to which later you can refer as
`BATTERY_PERCENTAGE_REMAINING`
(this particular example is from https://github.com/SmartThingsCommunity/SmartThingsPublic/blob/master/devicetypes/smartthings/leaksmart-water-sensor.src/leaksmart-water-sensor.groovy#L62 )

Later, you can compare them by using `zigbeeMap.clusterInt` instead of `zigbeeMap.cluster`.
Example of usage: https://github.com/SmartThingsCommunity/SmartThingsPublic/blob/a7c980340992ca37508d42a70f354b5fd7d30048/devicetypes/smartthings/zigbee-power-meter.src/zigbee-power-meter.groovy#L58

Reply:
updated magic numbers.  I think I got the changes you requested.